### PR TITLE
Issue #6081: Do not flush entity cache when creating a new node

### LINKED ISSUE #6081: Should node cache be emptied when creating a new node?
## Description of the bug

In the core node.entity.inc is [this line](https://github.com/backdrop/backdrop/blob/04c37cca201a952cbdb2911e1462ad6b84f2a90d/core/modules/node/node.entity.inc#L496):
```php
$this->resetCache($op == 'update' ? array($entity->{$this->idKey}): array());
```

when the `$op` is `'insert'`, it sends an empty array to `resetCache`, which results in it doing this:
```php
cache_flush('cache_entity_' . $this->entityType);
```
Is there a reason for it to need to do that?

It seems unnecessary, and if you have a page that is very complex (perhaps badly written) and slow to load when the cache is empty, then that page loads slowly after new nodes are created.

I would suggest that the line in node.entity.inc be rewritten to this:
```php
if($op == 'update') $this->resetCache(array($entity->{$this->idKey}));
```
So it doesn't clear the cache when inserting new nodes.

## Steps To Reproduce

You need to have a page that is very complex, that loads slowly if there's no cache, but that runs fast with a cache. Have a way to add new data that affects that page content. In my case a dialog opens on top with a form, the form submission is via ajax, the node is saved and the page content is reloaded via ajax. 

With core as it is now, that node save clears the cache and the reload takes 1 minute to complete. If I make the change I propose, the reload takes 1 second and the change is rendered (I don't need the cache to clear to see the change).

I'm afraid that I can't share my code or the data, and I don't have the time currently to spend creating an anonymous test set up for this.

## Actual behavior

As I said, the cache is cleared on a node save, and the page content reload takes one minute to complete.

## Expected behavior

I don't need the node cache to be cleared when a node is created.

## Additional information

This bug report is not to optimise my code, even though that is of course a good thing to do. I am questioning that decision to clear the entire node cache when a new node is created - why does that need to happen?

My site is in development, hosted locally on my own PC with WSL2, Ubuntu and Lando.

Of course my code could be optimised, and perhaps that's the solution. I'm migrating code from Drupal 7 that worked for years.

Backdrop CMS version: 1.24.0
Web server: Apache/2.4.54 (Debian)
PHP version: 7.4.33
Database server: MySQL 5.7.29

### PR DESCRIPTION
**Pull request by indigoxela**
Fixes https://github.com/backdrop/backdrop-issues/issues/6081

### CODE REVIEW

**Comment by herbdool** [comment]
Rebasing this PR branch may fix the phpcs errors.

File: core/modules/book/book.module [hunk 1]
```diff
@@ -662,6 +662,11 @@ function _book_update_outline(Node $node) {
         backdrop_static_reset('book_get_books');
       }
     }
+    // Get all nodes that are part of this book and flush their caches.
+    $affected_nids = db_query("SELECT nid FROM {book} WHERE bid = :bid", array(
+      ':bid' => $node->book['bid'],
+    ))->fetchAllAssoc('nid');
+    entity_get_controller('node')->resetCache(array_keys($affected_nids));
```
Comment (argiepiano) [inline core/modules/book/book.module:669]: This looks OK, but my concern about this is that `_book_update_outline()` is called multiple times when you delete a book page, once for each of the deleted book page children (which may mean a LOT of times if you delete the top-most node). Therefore we'd be running this query for each of the children (with the same resulting nids each time), and reseting the cache for the same nids over and over. While the approach here reuses code, it seems wasteful in that situation.

An alternative may be to create a new function (e.g. `_book_clear_cache_entity_node($node)`) that gets called from `book_node_insert()`, `book_node_update()` and `book_node_predelete()` only once.  This function would do the same as this code, take care of reseting the cache for all nids in the book, but it would avoid the unnecessary repetition resulting from deleting a book page.

Comment (indigoxela) [reply core/modules/book/book.module:669]: > ... is called multiple times when you delete a book page

Ah, that's a valid concern. Have to take a closer look re node deletion - in which case this call is useless. :+1:

Comment (indigoxela) [reply core/modules/book/book.module:669]: My idea was, that if a book gets deleted, we wouldn't end up here, as `empty($node->book['bid'])` and `if (menu_link_save...` should catch that. And then the outline isn't rebuilt. :thinking: Have to check that again, though.

Edit: it *is* rebuilt. See details below.

Comment (indigoxela) [reply core/modules/book/book.module:669]: Fancy Book module. :wink: 

@argiepiano The loop in book_node_predelete() was probably your concern, but that behaves correctly. Function _book_update_outline() runs once for every newly created "stub" book, if a top level node gets deleted.

Example:
```
- The book
- - Subpage 1
- - - Below subpage 1
- - Subpage 2
- - - Below subpage 2
```
When deleting "The book", two structures get created, one from "Subpage 1" with its children, one from "Subpage 2" with its children. So inside the loop the book ID is a different one and child nodes are different, too. No overlap as far as I can see. Or am I missing something?

Comment (argiepiano) [reply core/modules/book/book.module:669]: Right! This looks good then!

**Review by klonos — APPROVED** [review]
LGTM 👍🏼

**Comment by quicksketch** [comment]
Looks good! Merged into 1.x and 1.24.x.
